### PR TITLE
First implementation of third party templates

### PR DESCRIPTION
**Pull request by kunaldeo**
Here is the complete working implementation of third-party template system for stagehand.
To install a third party package:
```stagehand --git <Git Repo URL>```
You can use this example third party template at  https://github.com/kunaldeo/stagehandtrialtemp. Template has the following structure:
```
├── main.dart
├── pubspec.yaml
└── src
    └── stagehandtrialtemp_data.dart
```
This pull request merges all the commits in PR https://github.com/google/stagehand/pull/202 . Please see it for earlier discussions.

### CODE REVIEW

**Comment by kunaldeo** [comment]
@sethladd I have merged all the changes into this.

**Comment by sethladd** [comment]
gotcha, thanks!

**Comment by sethladd** [comment]
cc @devoncarew

**Comment by sethladd** [comment]
We should probably update the README with instructions on how to create a plugin, and how to consume a plugin.

**Comment by kunaldeo** [comment]
I am working on a tools/grind.dart file to help create the plugin. Once that is done we can update the README.

File: lib/src/cli_app.dart [hunk 1]
```diff
@@ -122,6 +122,45 @@ class CliApp {
...
+      String templateLocation = path.join(stagehandDataDir, repoName);
+
+      if(io.FileSystemEntity.isDirectorySync(templateLocation)) {
+        logger.stderr("Template already installed. If you want to update the template use the update command\n");
+        _usage(argParser);
+        //TODO: Implement template update command
+        return new Future.error(new ArgError('Template already installed.'));
+      }
+
+      logger.stdout("Clonning " + repoName);
```
Comment (andreasolund) [inline lib/src/cli_app.dart]: Should be "Cloning"

**Comment by kunaldeo** [comment]
Added instructions on how to create and consume a third party template. 

Bootstrap Template: [emptyStagehandTemplate](https://github.com/kunaldeo/emptyStagehandTemplate), This will give developers starting point to build new templates. 
Fully Built Sample Template : [polymerContacts](https://github.com/kunaldeo/polymerContacts)

**Comment by sethladd** [comment]
Thanks very much for the pull request! It spawned a lot of discussions here around the water cooler today, mainly about how best to enable a thriving ecosystem of templates, the differences between declarative and imperative solutions, and maintenance costs.

We're impressed by the isolates-based plugins approach. For systems that require imperative plugins (plugins that have to execute code), it's a solid approach. Isolates give us safety from third-party code, and a way to load the code. The plugin package's API makes it easier to do it.

We asked ourselves: "What do the plugins need to execute?" and we couldn't come up with a solid answer. We know a template needs to ask certain questions and display certain answers, but does it need to execute code? Can we get away with _declarative_ templates?

We then asked ourselves, "If we don't need a plugin to execute code, do we want to support and API for those plugins?" The answer to that was quickly: if we can avoid providing a Dart API for the plugins, that's less for us (and plugin maintainers) to worry about, version, etc.

Ultimately, in order to strive for a simple system that asks very little of our template authors, we couldn't find a really compelling reason to go the plugin/isolates/code approach. A design that uses pub itself to distribute and upgrade templates, and a simple declarative approach to collecting info, seems like the most efficient way to accelerate the growth of an ecosystem of plugins for stagehand.

We are extremely appreciative of your hard work, and we don't think we could have found a solution without your pull request. In the end, we feel we have a simpler way forward that doesn't require isolates, and thus we unfortunately aren't going to merge this PR. Devon will follow up with a design that hopefully satisfies all of our use cases, so we ensure we're allowing template authors to do all the great stuff that you envisioned.

Again, even though we don't plan to merge this, we are extremely appreciative of your time and energy, and want to thank you for being part of this project.

**Comment by kwalrath** [comment]
Here are my notes on the design from talking to Devon.

A template could be in a pub package or a github repo. Say it's a pub package. You’d pub global activate that package. Stagehand would enumerate all the activated packages and look for a file like stagehand.yaml in them. Then it needs to load stagehand packages. That was what stopped them before, but now that the resource API is available (in 1.14), loading stagehand packages should be much easier.

For an example of using the resource API, see [this recent dartdoc change](https://github.com/dart-lang/dartdoc/commit/d5f59376d7cc8c8b97d2b9b312b63667d0265c25).

**Pull request closed without merging**